### ISSUE TITLE
[css-conditional-4] Need a way in CSS to test for support of HTML features

### ISSUE THREAD

**Issue by jensimmons** [w3c/csswg-drafts#9746]
There should be a way to feature detect support for an HTML attribute, HTML element and other things in HTML. 

For instance, the `switch` attribute is new to HTML. It turns a checkbox form control into a native switch, in a progressively enhanced fashion.

Authors will likely want to style their switch with custom styles, while still supporting the checkbox fallback. If they apply styles without writing conditional code, the checkbox is ruined. 

This is one idea of how it could work...
```
@supports selector(input[switch]) {
 /* Would like to put custom switch styles here, so they only apply conditionally. */
}
```
...but that doesn't work... because it always returns true. The selector is understood by CSS, even when it's not supported in HTML. 

How can we provide a mechanism to CSS for testing support of HTML features?

Here's a demo of switch, with custom styles that should be applied conditionally. 
https://codepen.io/jensimmons/pen/eYXYZWr?editors=1100

**Comment by hober** [comment]
At least in the case of `<input type=checkbox switch>`, I think the more pressing concern from CSS is to be able to test for support of the `::track` and `::thumb` pseudo-elements. Which is maybe covered by `@supports selector()`?

**Comment by hober** [comment]
Put another way, the reason you want feature detection from CSS is so that you can style the thing on one way instead of in another way. Concretely:

* Write styles that apply to checkboxes but not switches in browsers that support just checkboxes as well as browsers that support both
* Write styles that apply to switches in browsers that support them and don't get applied at all in browsers that don't support switches)

The first one is just `input:not([switch])`, I think.
The second is really interesting. @nt1m tells me `:has()` doesn't support pseudo-elements, which is a real bummer, because `input[type=checkbox][switch]:has(::track)` would be the obvious and nice solution to this if it worked.

**Comment by fantasai** [comment]
Yeah, I think this would be a good idea. Could even re-use selector syntax, but it would be a different function, e.g. `html()` or `markup()` or something. E.g.

-  `HTML(rb)` indicates the UA supports the `rb` element
- `HTML([dir])` indicates some basic level of support for the `dir` attribute (HTML4-level support would be adequate to return true, because `auto` hadn't been invented yet)
- `HTML([dir=auto])` indicates support for `dir=auto`
- `HTML(input[type=color])` indicates support for `<input type=color>`

CSS tracks whether features are supported by whether they're successfully parsed, and you probably can't do that here, though so it would require the UA maintaining some kind of registry of what it supports.

**Comment by SebastianZ** [comment]
I really like @fantasai's proposal. I'd just note to think of SVG and MathML, as well. So either call the function `markup()` and let it cover all types of markup. Or, to avoid name clashes and allow more precise checks, let the `markup()` function take keywords to differenciate them or introduce separate functions `html()`, `svg()`, and `mathml()` for them.

Sebastian

**Comment by nt1m** [comment]
The problem with the above proposal is that a registry of features for each language could go stale, and there's no good way from code to enforce that registry from being updated.

@annevk suggested linked with IDL, which I think is a much better idea, since IDL _needs_ to be updated in order to expose a new web feature.

`@supports idl(HTMLInputElement.prototype.switch)` or such. (syntax to be bikeshed)

**Comment by SebastianZ** [comment]
While IDL is good in regard of requiring to be updated, not all elements have an IDL associated to them, like e.g. the `<ruby>` element.

Therefore, I'd vote for introducing a similar registry for markup, as that would not only allow CSS to check whether something is supported but also have other benefits. E.g. having such a registry would allow DevTools to add autocompletion for tag and attribute names without having to maintain them by themselves. And it also makes markup feature detection generally easier or even possible, which is useful to authors and presumably also for WPT.

Though if such a registry is not possible for some reason and we have to base the checks on IDL, the syntax should be made more CSSy and also _not_ include the abbreviation "IDL" as that's rather a spec term than an author-facing one. It could actually still be mapped to what @fantasai and I proposed earlier, because all IDLs follow a specific naming syntax. The names have the type as prefix, followed by their name, followed by the suffix "Element". And the IDLs also define the attributes. So you could internally translate `html(input[switch])` to `HTMLInputElement.prototype.switch` or `svg(clipPath)` to `SVGClipPathElement`.
But again, it would be weird if you _can_ test for `html(details)` but _not_ for `html(summary)` just because there's no `HTMLSummaryElement` interface for it.

Sebastian

**Comment by nt1m** [comment]
You mentioned DevTools are an example of registry, I do want to say this is a counter example, this isn't something WebKit does a great job at updating well.

I agree the syntax should be as friendly as possible, but not having to maintain a separate registry is as equally as important IMO.

**Comment by nt1m** [comment]
I guess engines have an atom string registry for both tagnames / attributes, which can be re-used, but I don't think it's well standardized or consistent across browsers (whereas IDL is more well standardized)

**Comment by SebastianZ** [comment]
> You mentioned DevTools are an example of registry, I do want to say this is a counter example, this isn't something WebKit does a great job at updating well.

No, what I meant is that DevTools currently have to maintain their own registry for that. I advocate for a central registry of HTML elements and attributes _in the engine_. So DevTools could make use of that instead. And that central registry would also cater the feature detection used in CSS, which is discussed here.

> I guess engines have an atom string registry for both tagnames / attributes, which can be re-used, but I don't think it's well standardized or consistent across browsers (whereas IDL is better standardized)

_That_ is my point. To allow CSS to detect HTML features, the engines need to have some kind of registry it can use and check against. Additional standardization across user agents doesn't seem to be necessary, engines only need to ensure their registries are consistent and kept up-to-date.

Sebastian

**Comment by Loirooriol** [comment]
I share Tim's concern about a registry. It's basically the same problem as in https://github.com/w3c/csswg-drafts/issues/3559#issuecomment-458734231, so I guess @tabatkins is also opposed to a registry.

**Comment by nt1m** [comment]
> That is my point. To allow CSS to detect HTML features, the engines need to have some kind of registry it can use and check against. Additional standardization across user agents don't seem to be necessary, engines only need to ensure their registries are consistent and kept up-to-date.

If engines already had a registry that they are already using I would be happy to leverage it, but the current registries that are available are not great. I still stand by my point that any mechanism used here should be required to be kept up to date by design.

Unfortunately, currently available atom string registries of attribute names / tag names are too vague. E.g. once `switch` gets added to the attribute registry, then `h1[switch]` will also return true. This is why I still think IDL is the best mechanism to leverage even if not extensive.

**Comment by SebastianZ** [comment]
> > That is my point. To allow CSS to detect HTML features, the engines need to have some kind of registry it can use and check against. Additional standardization across user agents don't seem to be necessary, engines only need to ensure their registries are consistent and kept up-to-date.
> 
> If engines already had a registry that they are already using I would be happy to leverage it, but the current registries that are available are not great.

I obviously don't understand enough about the internals of the engines. Though I am wondering why engines e.g. have a consistent registry for CSS properties but not for HTML elements and attributes they can handle.

And a way to test for those features would pave the way for WPTs to test for them. And by having tests for all the different HTML features, we avoid the issue of implementations becoming outdated and push them towards introducing proper registries, so they don't have to manually maintain them.

> I still stand by my point that any mechanism used here should be required to be kept up to date by design.

@nt1m Which other mechanism than the above could you imagine that allows for a comprehensive solution?
Should the HTML spec. maintain machine readable definitions for all HTML elements and attributes?

Sebastian

**Comment by SebastianZ** [comment]
> I share Tim's concern about a registry. It's basically the same problem as in [#3559 (comment)](https://github.com/w3c/csswg-drafts/issues/3559#issuecomment-458734231), so I guess @tabatkins is also opposed to a registry.

It's not exactly the same. There, the issue is that implementations would have to maintain contextual information, i.e. whether one property is interpreted in combination with another.
Here, the check aligns with the idea of "Does it parse?", as Tab expressed it. For `@supports (grid-template-columns: subgrid)` the engine needs to be able to parse the property and the given value. So the property name and its syntax need to be registered. For `@supports html(input[switch])` it's similar.
And if it's not close enough to the "Does it parse?" principle yet, how about the following syntax?

```css
@supports html(<input type="checkbox" switch>) { … }
```

Sebastian

**Comment by nt1m** [comment]
> I obviously don't understand enough about the internals of the engines. Though I am wondering why engines e.g. have a consistent registry for CSS properties but not for HTML elements and attributes they can handle.

This is because the HTML parser allows arbitrary name / value pairs of attributes, unlike CSS where invalid props won't actually parse.

WebKit has:
1. a list of attributes it handles: https://searchfox.org/wubkat/source/Source/WebCore/html/HTMLAttributeNames.in
2. a list of tagnames it handles: https://searchfox.org/wubkat/source/Source/WebCore/html/HTMLTagNames.in

(These can go stale if one gets removed, but accidentally stay in the list)

Potentially a mix of 2 + IDL lookup could work, but then any attributes that don't have an IDL equivalent might not work (and sometimes the IDL name doesn't match the attribute name too).

> And a way to test for those features would pave the way for WPTs to test for them. And by having tests for all the different HTML features, we avoid the issue of implementations becoming outdated and push them towards introducing proper registries, so they don't have to manually maintain them.

I wouldn't fully rely on WPT to ensure this is up to date. It would really have to be something that's mandatory to update when you add a new feature.

**Comment by Loirooriol** [comment]
@SebastianZ There is no registry for the supported CSS properties, browsers just check whether the parser accepts that. There is no promise of functional support, e.g. `@supports (gap: 1px)` doesn't imply that flex layout will use it, and `@supports (content: "+")` doesn't imply that it will work on a `::marker`.

But in this case the HTML parser accepts arbitrary elements and attributes, so it's not as easy as checking whether the parser accepts it.

> While IDL is good in regard of requiring to be updated, not all elements have an IDL associated to them, like e.g. the `<ruby>` element.

Not a big deal, the interface is `HTMLElement`, "support" for elements could be defined as getting an interface different than `HTMLUnknownElement`. Attribute "support" is trickier, since validation typically happens on IDL attributes, not content attributes.

> Here, the check aligns with the idea of "Does it parse?"

The registry idea does not align with that. And `<input type="checkbox" switch>` parses perfectly fine even without support for `switch`.

**Comment by Crissov** [comment]
Don’t authors usually just want to check whether the browser’s internal stylesheet for HTML (or SVG or MathML) contains any rule for the element or attribute they are querying?

**Comment by Loirooriol** [comment]
@Crissov Not necessarily, lots of things are supported with internal magic instead of CSS, e.g. I'm not finding any CSS targeting the attribute `maxlength` on Gecko, Blink nor WebKit, but they support it on `<input>`.

**Comment by tabatkins** [comment]
>  @nt1m tells me :has() doesn't support pseudo-elements

Not quite; the spec *allows* for `:has()` to take certain pseudo-elements (the [:has-allowed pseudo-elements](https://drafts.csswg.org/selectors/#relational)), but doesn't define any of the existing ones to be in that set.

The basic requirement for a pseudo to be :has-allowed is just that it's not dependent on styling, so you can't introduce cycles with it. Would you be able to shut off the switch pseudos, with `appearance` or something? If so, they can't be :has-allowed; if not, then they could be.

(The hidden extra requirement is that they're not "always there" either, like ::before, because testing for them would be useless in that case.)

----------

Returning to the main topic, yeah, as much as possible we don't want to maintain registries, and want to instead rely on parsing to give us a correct answer automatically. We can probably pull that off with tagnames (based on IDL class, as @Loirooriol said), and likely with input types (they have their own behavior validator that recognizes valid names), but not with attributes in general. Every attribute we add would need special handling.

**Comment by annevk** [comment]
You could base it off IDL members, but I'm somewhat hesitant to build IDL testing into CSS.

**Comment by francescorn** [comment]
Authors can use the :is() pseudo-class possibly such as:
@supports (:is(input[switch])) {
  input[switch] {
    appearance: switch;
  }
}

**Comment by bramus** [comment]
In https://bsky.app/profile/tfedder.de/post/3mu3t7yxqzs2g a new use-case was mentioned: checking for support for the `request-close` invoker command. IUC `html()` would solve this use-case: `html(button[command="request-close"])`
